File: linked lists/86. Partition List/partition.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def partition(self, head: Optional[ListNode], x: int) -> Optional[ListNode]:
        left, right = ListNode(), ListNode() # Dummy node
        last_left, last_right = left, right
        curr = head

        while curr:
            if curr.val < x:
                last_left.next = curr
                last_left = last_left.next
            else:
                last_right.next = curr
                last_right = last_right.next
            curr = curr.next
        
        last_right.next = None
        last_left.next = right.next # Skip dummy node
        return left.next

File: linked lists/86. Partition List/partition.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} x
 * @return {ListNode}
 */
var partition = function(head, x) {
    let left = new ListNode(), right = new ListNode();
    let lastLeft = left, lastRight = right, curr = head;

    while (curr) {
        if (curr.val < x) {
            lastLeft.next = curr;
            lastLeft = lastLeft.next;
        } else {
            lastRight.next = curr;
            lastRight = lastRight.next;
        }
        curr = curr.next;
    }

    lastRight.next = null;
    lastLeft.next = right.next;
    return left.next;
};